saw-core/src/Verifier/SAW/Simulator/TermModel.hs

```diff
@@ -0,0 +1,1092 @@
+              a  <- scBitvector sc (wordWidth x)
+              tm' <- scIte sc a tm x' y'
+              pure (Left (wordWidth x, tm'))
+
+  , Prims.bpMuxInt = \c x y ->
+       case c of
+         Right b -> if b then pure x else pure y
+         Left tm ->
+           do x' <- intTerm sc x
+              y' <- intTerm sc y
+              a  <- scIntegerType sc
+              Left <$> scIte sc a tm x' y'
+
+  , Prims.bpMuxExtra = \tp c x y ->
+       case c of
+         Right b -> if b then pure x else pure y
+         Left tm ->
+           do x' <- extraTerm x
+              y' <- extraTerm y
+              a  <- readBackTValue sc cfg tp
+              VExtraTerm tp <$> scIte sc a tm x' y'
+
+    -- Booleans
+  , Prims.bpTrue   = Right True
+  , Prims.bpFalse  = Right False
+  , Prims.bpNot    = unOp sc scNot not
+  , Prims.bpAnd    = binOp sc scBool scAnd (&&)
+  , Prims.bpOr     = binOp sc scBool scOr  (||)
+  , Prims.bpXor    = binOp sc scBool scXor (/=)
+  , Prims.bpBoolEq = binOp sc scBool scBoolEq (==)
+
+    -- Bitvector logical
+  , Prims.bpBvNot  = bvUnOp  sc scBvNot (Prim.bvNot undefined)
+  , Prims.bpBvAnd  = bvBinOp sc scBvAnd (Prim.bvAnd undefined)
+  , Prims.bpBvOr   = bvBinOp sc scBvOr  (Prim.bvOr  undefined)
+  , Prims.bpBvXor  = bvBinOp sc scBvXor (Prim.bvXor undefined)
+
+    -- Bitvector arithmetic
+  , Prims.bpBvNeg  = bvUnOp  sc scBvNeg  (Prim.bvNeg  undefined)
+  , Prims.bpBvAdd  = bvBinOp sc scBvAdd  (Prim.bvAdd  undefined)
+  , Prims.bpBvSub  = bvBinOp sc scBvSub  (Prim.bvSub  undefined)
+  , Prims.bpBvMul  = bvBinOp sc scBvMul  (Prim.bvMul  undefined)
+  , Prims.bpBvUDiv = bvDivOp sc scBvUDiv (Prim.bvUDiv undefined)
+  , Prims.bpBvURem = bvDivOp sc scBvURem (Prim.bvURem undefined)
+  , Prims.bpBvSDiv = bvDivOp sc scBvSDiv (Prim.bvSDiv undefined)
+  , Prims.bpBvSRem = bvDivOp sc scBvSRem (Prim.bvSRem undefined)
+  , Prims.bpBvLg2  = bvUnOp  sc scBvLg2  Prim.bvLg2
+
+    -- Bitvector comparisons
+  , Prims.bpBvEq   = bvCmpOp sc scBvEq  (Prim.bvEq  undefined)
+  , Prims.bpBvsle  = bvCmpOp sc scBvSLe (Prim.bvsle undefined)
+  , Prims.bpBvslt  = bvCmpOp sc scBvSLt (Prim.bvslt undefined)
+  , Prims.bpBvule  = bvCmpOp sc scBvULe (Prim.bvule undefined)
+  , Prims.bpBvult  = bvCmpOp sc scBvULt (Prim.bvult undefined)
+  , Prims.bpBvsge  = bvCmpOp sc scBvSGe (Prim.bvsge undefined)
+  , Prims.bpBvsgt  = bvCmpOp sc scBvSGt (Prim.bvsgt undefined)
+  , Prims.bpBvuge  = bvCmpOp sc scBvUGe (Prim.bvuge undefined)
+  , Prims.bpBvugt  = bvCmpOp sc scBvUGt (Prim.bvugt undefined)
+
+    -- Bitvector shift/rotate
+  , Prims.bpBvShlInt = \c x amt ->
+      case (c, x) of
+        (Right c', Right x') -> pure . Right $! Prim.bvShiftL c' x' amt
+        _ -> do let n = wordWidth x
+                n'  <- scNat sc n
+                a   <- scBoolType sc
+                ctm <- boolTerm sc c
+                xtm <- wordTerm sc x
+                amttm <- scNat sc (fromInteger amt) -- TODO, should probably be a Natural?
```

Comment:
This would require a change to the simulator API, right?

Reply:
Yes, it will be a pretty big refactoring, and ultimately bumps into the same `Vector` issue as I mentioned before.